src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php

```diff
@@ -300,6 +289,444 @@ public static function analyze(
         return false;
     }
 
+    public static function castIntAttempt(
+        StatementsAnalyzer $statements_analyzer,
+        Context $context,
+        Union $stmt_type,
+        PhpParser\Node\Expr $stmt,
+        bool $explicit_cast = false
+    ): Union {
+        $codebase = $statements_analyzer->getCodebase();
+
+        $risky_cast = [];
+        $invalid_casts = [];
+        $valid_ints = [];
+        $castable_types = [];
+
+        $atomic_types = $stmt_type->getAtomicTypes();
+
+        $parent_nodes = [];
+
+        if ($statements_analyzer->data_flow_graph instanceof VariableUseGraph) {
+            $parent_nodes = $stmt_type->parent_nodes;
+        }
+
+        while ($atomic_types) {
+            $atomic_type = array_pop($atomic_types);
+
+            if ($atomic_type instanceof TInt) {
+                $valid_ints[] = $atomic_type;
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TFloat) {
+                if ($atomic_type instanceof TLiteralFloat) {
+                    $valid_ints[] = new TLiteralInt((int) $atomic_type->value);
+                } else {
+                    $castable_types[] = new TInt();
+                }
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TString) {
+                if ($atomic_type instanceof TLiteralString) {
+                    $valid_ints[] = new TLiteralInt((int) $atomic_type->value);
+                } elseif ($atomic_type instanceof TNumericString) {
+                    $castable_types[] = new TInt();
+                } else {
+                    // any normal string is technically $valid_int[] = new TLiteralInt(0);
+                    // however we cannot be certain that it's not inferred, therefore less strict
+                    $castable_types[] = new TInt();
+                }
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TNull || $atomic_type instanceof TFalse) {
+                $valid_ints[] = new TLiteralInt(0);
+                continue;
+            }
+
+            if ($atomic_type instanceof TBool) {
+                $valid_ints[] = new TLiteralInt(1);
+                continue;
+            }
+
+            // could be invalid, but allow it, as it is allowed for TString below too
+            if ($atomic_type instanceof TMixed
```

Comment:
@orklah What do you think about [this behavior](https://psalm.dev/r/170b48f07d)? Leave it as is to avoid a bunch of `mixed` issues?

It annoys me a bit because I sometimes have to use `mixed` to represent an unknown value that's able to be anything, but I still want to be warned if I do anything invalid with it. Unfortunately for me Psalm likes to ignore a lot of problems with `mixed` types because otherwise untyped codebases would have a ridiculous number of errors.

Reply:
Well, in my mind at least, whenever I see an explicit cast, I tend to assume the developper knew what he was doing. I don't know if this hunch is correct and verified in the wild, but I would be personnally annoyed if Psalm were to report me such cases (If the projects I analyze were at level 1 at least!).

Reply:
I never like to assume the developer knows what they're doing! :stuck_out_tongue_closed_eyes:

But yeah, it fits with all the other special exceptions we have for `mixed`. Maybe I'll have to start using `object|resource|array|string|int|float|bool|null` instead, despite the verbosity.

Reply:
I was going to say that Psalm would revert that to mixed [like we do with scalar](https://github.com/vimeo/psalm/blob/b4fdc3e326fea865de49827f723a2703bb18617a/src/Psalm/Internal/Type/TypeCombiner.php#L334) but it doesn't seem so!